br/pkg/stream/stream_metas.go

```diff
@@ -408,3 +321,956 @@ func ReplaceMetadata(meta *backuppb.Metadata, filegroups []*backuppb.DataFileGro
+		result = Migrations{
+			Base:   new(pb.Migration),
+			Layers: collected.Item,
+		}
+	}
+	return result, nil
+}
+
+func (m MigrationExt) DryRun(f func(MigrationExt)) []storage.Effect {
+	batchSelf := MigrationExt{
+		s:      storage.Batch(m.s),
+		prefix: m.prefix,
+		Hooks:  m.Hooks,
+	}
+	f(batchSelf)
+	return batchSelf.s.(*storage.Batched).ReadOnlyEffects()
+}
+
+func (m MigrationExt) AppendMigration(ctx context.Context, mig *pb.Migration) (int, error) {
+	migs, err := m.Load(ctx)
+	if err != nil {
+		return 0, err
+	}
+	newSN := migs.Layers[len(migs.Layers)-1].SeqNum + 1
+	name := path.Join(migrationPrefix, nameOf(mig, newSN))
+	data, err := mig.Marshal()
+	if err != nil {
+		return 0, errors.Annotatef(err, "failed to encode the migration %s", mig)
+	}
+	return newSN, m.s.WriteFile(ctx, name, data)
+}
+
+// MergeTo merges migrations from the BASE in the live migrations until the specified sequence number.
+func (migs Migrations) MergeTo(seq int) *pb.Migration {
+	return migs.MergeToBy(seq, MergeMigrations)
+}
+
+func (migs Migrations) MergeToBy(seq int, merge func(m1, m2 *pb.Migration) *pb.Migration) *pb.Migration {
+	newBase := migs.Base
+	for _, mig := range migs.Layers {
+		if mig.SeqNum > seq {
+			return newBase
+		}
+		newBase = merge(newBase, &mig.Content)
+	}
+	return newBase
+}
+
+type mergeAndMigrateToConfig struct {
+	interactiveCheck       func(context.Context, *pb.Migration) bool
+	alwaysRunTruncate      bool
+	appendPhantomMigration []pb.Migration
+}
+
+type MergeAndMigrateToOpt func(*mergeAndMigrateToConfig)
+
+func MMOptInteractiveCheck(f func(context.Context, *pb.Migration) bool) MergeAndMigrateToOpt {
+	return func(c *mergeAndMigrateToConfig) {
+		c.interactiveCheck = f
+	}
+}
+
+// MMOptAlwaysRunTruncate forces the merge and migrate to always run the truncating.
+// If not set, when the `truncated-to` wasn'd modified, truncating will be skipped.
+// This is necessary because truncating, even a no-op, requires a full scan over metadatas for now.
+// This will be useful for retrying failed truncations.
+func MMOptAlwaysRunTruncate() MergeAndMigrateToOpt {
+	return func(c *mergeAndMigrateToConfig) {
+		c.alwaysRunTruncate = true
+	}
+}
+
+// MMOptAppendPhantomMigration appends a phantom migration to the merge and migrate to operation.
+// The phantom migration will be persised to BASE during executing.
+// We call it a "phantom" because it wasn't persisted.
+// When the target version isn't the latest version, the order of migrating may be broken.
+// Carefully use this in that case.
+func MMOptAppendPhantomMigration(migs ...pb.Migration) MergeAndMigrateToOpt {
+	return func(c *mergeAndMigrateToConfig) {
+		c.appendPhantomMigration = append(c.appendPhantomMigration, migs...)
+	}
+}
+
+// MergeAndMigrateTo will merge the migrations from BASE until the specified SN, then migrate to it.
+// Finally it writes the new BASE and remove stale migrations from the storage.
+func (m MigrationExt) MergeAndMigrateTo(
```

Comment:
I found there are many functions only used in unit test, so I prefer to move these functions to another file with suffix `_test.go` to read and check codes for production more clearly.